Vue CKEditor 4 Component implementation

engineering-this opened the pull request:
# Early PR of CKEditor4-Vue Integration

File structure, utils, scripts, configuration files, etc. borrowed from https://github.com/ckeditor/ckeditor5-vue

## Source files
- `/src/ckeditor.js` - CKEditor component, nothing fancy there
- `/src/plugin.js` - Vue plugin, which just wraps the component

## Sample
To run sample you need to build first, by `npm run build`, then you can just open `sample/index.html` in your browser. I've put all the components templates in the `html` files and all logic in `main.js`. The main purpose of sample for now is manual testing. After some refactoring and polishing it might be used as [docs](https://ckeditor.com/docs/ckeditor4/latest/examples/index.html) example.

### Notes for sample
- Multiple components in one file is Vue anti-pattern, however I did it for now to keep it as simple as possible.
- Before including `load-script` and support of `editorUrl` property component could be used directly from source by importing it as ES6 Module and using `type="module"` attribute on script which included it. If for any reason we wan't to allow using it from source again, than `load-script` should be included directly in our repo instead of a dependency.

## Tests
Run with `npm run tests`.
To run tests for Firefox `npm run tests -- -b Firefox`.

For other flags look [here](https://github.com/ckeditor/ckeditor4-vue/blob/839152d7ca5a69d6da03a53cb175c0d7a24c0bbd/scripts/utils/getkarmaconfig.js#L237-L264) - all borrowed from CKE5-Vue.

Just like in `ckeditor5-vue` I made test of two kind:
- Unit tests, which uses mocked editor: `/tests/component.js`, `/tests/plugin.js`, `tests/helpers.js`
- Integration tests, which uses actual editor: `/tests/integration.js`

## Including CKEditor
Just like in other CKE4 frameworks integration component will detect if `CKEDITOR` namespace is present, if not it will load from CDN. Also `ckeditor.js` url can be specified by `editorUrl` prop.

## Component features
Component can be initialized by using ckeditor tag in parent's component template:
```<ckeditor></ckeditor>```

### Supported optional property attributes:
- `value` - editor data, it needs to be named like it, so `v-model` can be used for two-way binding. We could add an alias `data`, but having two attributes doing same thing is not the best thing to do.
- `type` - editor type, `classic` or `inline`.
- `editorUrl` - link to `ckeditor.js` file, needed only if `CKEDITOR` namespace is not loaded other way.
- `config` - `CKEDITOR.config` object.
- `tagName` - Tag name for an element which is to be used for editor creation.
- `readOnly` - If editor should be initially read only.
- `v-model` - for two way data binding.

### Supported optional event attributes:
- `@input` - corresponds with `editor#change` event, needed for `v-model`.
- `@focus` - `editor#focus` event.
- `@blur` - `editor#blur` event.
- `@ready` - when component is ready.
- `@destroy` - when component is destroyed.

## Issues
- When editor is unmounted before it is fully initialised - it throws errors.
This issue was fixed in Angular by initializing editor outside of component, and then moving it to right place.
- When component with classic editor is destroyed after iframe is removed from DOM CKEditor shows error `editor-destroy-iframe`.
This was _fixed_ in Angular by removing option to use classic editor...

Vue router offers guards out of the box, so I've added method `destroyEditor` which will return promise, which resolves once editor is destroyed. In samples I've declared guards that delay destroying component until editor is destroyed. This fixes both mentioned above issues, we should advocate using guards like this in our docs.

Another case which reproduces above issues is when component is unmounted by `v-if`. I've managed to add guard-like logic in samples for `v-if`, which also does the trick.

Note: You might find some leftovers here and there.

## Todo

Actually it's hard to list any todos here, it works nice, scripts for building from CKE5 works, so apart from so general stuff like updating `README` and `PACKAGE.json` I don't see anything. But let's see what flaws will be found during review.

## Helpful links:

https://vuejs.org/v2/guide/
https://learn-vuejs.github.io/vue-patterns/patterns/

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,157 @@
+						this.instance.destroy();
+						this.instance = null;
+					} else {
+						res();
```
It's a bit of cheating here. If destroy is called with no instance, isn't the signal something goes wrong with component lifecycle?

engineering-this replied:
Component itself doesn't trigger such case, but Editor can be destroyed directly via API, third party plugin. Component is just a simply wrapper it can't be designed for every possible use case. IMO framework integration shouldn't limit Editors flexibility.

engineering-this replied:
That's just not possible. Some leftover from debugging tests.

engineering-this commented on src/helpers.js:
```diff
@@ -0,0 +1,33 @@
+let promise;
+
+export function getEditorNamespace( editorURL ) {
+	if ( editorURL.length < 1 ) {
```
>What if I don't care about URL because I know that I already attached CKEditor externally? Will I need to provide some fake URL to get around this condition?

In this case default value will be provided by framework, yet if `CKEDITOR` namespace is present no scripts are loaded.

I'll slightly rework it tho, so it won't throw when `editorUrl` is overridden with empty string, but `CKEDITOR` namespace is present.

> we should throw error where there is actually no CKEditor instance loaded by loadScript function?

👍

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,157 @@
+		},
+		editorUrl: {
+			type: String,
+			default: 'https://cdn.ckeditor.com/4.11.4/standard/ckeditor.js'
```
Existing `CKEDITOR` namespace takes precedence over default, or even passed url. With default value the component will work out of the box, also this is consistent with both React and Angular integrations.

jacekbogdanski replied:
I see how it's working. I don't understand why we need multiple ways to include CKEditor4 instance. Such hardcoded URL seems redundant and requires maintaining.

engineering-this replied:
I don't have strong opinion here. But maybe it's worth to have bigger discussion here. If we decide to drop hardcoding `editorUrl`, then we could do the same in other frameworks, they are still in `beta` status, so it's not so unexpected to have breaking changes.

Comandeer replied:
I would leave URL intact, as it allows lazy loading CKE just before creating editor's instance. Global `CKEditor` must be already bundled/loaded when someone attempts to create editor.

TBH I'm thinking about resolving this issue on entirely different level – introduce `latest` "version" to our CDN:

```javascript
'https://cdn.ckeditor.com/latest/standard/ckeditor.js'
```

OTOH we have already automated it in React, so copying and modifying script for Vue shouldn't be much work.

wwalc replied:
Using "latest" in the URL and a CDN is kinda incompatible. You'd have to set a very very short caching level (and CDN's are rather not for that) in order to minimize the negative consequences of it. Suppose we set `max-age=3600, must-revalidate`. If we change the CKEditor library under the same URL, then there is a non-zero chance, that within the next hour a user who already loaded CKEditor 4 (`ckeditor.js`),will face issues caused by additional external resources were not loaded. Imagine e.g. a scenario where CKEditor ver. N tries to load e.g. a dialog window or a translation coming from version N+1. In some situations may lead to a broken editor.

wwalc replied:
Well, the issue will actually occur always, regardless if the CDN is used or not, as long as you are not using a versioned folder.

wwalc replied:
The cache of a CDN will eventually cause additional issues if someone creates a plugin which is using only the newest version of a plugin and count on people having the latest version of CKEditor served from a CDN. While some of them may have still cached the older version. Even cache invalidation on edge locations of a CDN does not help (in reality).

f1ames replied:
I would be for keeping the `editorUrl` as it makes using integration easier. You just install one package and are ready to go. And it assures the user that integration works with the given editor version (which is provided by default).

Without default one it requires additional checking which CKEditor version you should/could use with the integration (well, ATM all version from like few years back should work fine, but user doesn't have to know or want to think about that). You can change version manually, but then if something breaks it's obvious that it's because of editor version change.

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,157 @@
+		return this.readOnly === null
+			? this.instance.readOnly
+			: this.readOnly;
+	},
```
Using default value would break `config.readOnly`. Component attribute should take precedence, but still it should be possible to use config option. It is common that some project are rewritten with new frameworks, but they might use existing editor config.

jacekbogdanski replied:
Ok, I see the issue now. I was thinking about updating `this.config` but not sure if it will make it more readable. Something like:
```Object.assign( {}, this.config, { readOnly: this.readOnly } )```
to avoid changing user object. In that case you could just use simple `readonly() { return this.instance.readOnly } )` also without setting this mode later on `instanceReady`.

engineering-this replied:
Slightly changing your idea, and we can get rid of `readOnly` getter. It looks like readability will be improved even further.

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,157 @@
+				} else {
+					this.$_setUpEditorEvents();
+					this.$emit( 'ready', editor );
+				}
```
`editor#change` is used to emit `component#input`, as data can be updated by ACF, we want to emit proper event. Firing `component#input` manually wouldn't be consisted with how normally it is fired with an original `change` evt passed as an param.

The `null` path can be omitted, as default value is ''.

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,157 @@
+		},
+
+		destroyEditor() {
+			if ( this.instance ) {
```
I made it public, so it can be used by router guards. However as https://github.com/ckeditor/ckeditor-dev/issues/3115 is going to be fixed, we won't need to expose this function as public.

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,157 @@
+	},
+
+	beforeCreate() {
+		this.$_ready = new Promise( res => {
```
With https://github.com/ckeditor/ckeditor-dev/issues/3115 we will be able to get rid of this.

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,157 @@
+
+	watch: {
+		value( val ) {
+			if ( this.instance && this.instance.getData() !== val ) {
```
`readOnly` should be validated too. It can be updated before instance is created, it would lead to an error.

jacekbogdanski replied:
It seems for me like developer error than an implementation issue. What's a use case for setting data / readonly mode before the editor is initialized? If it can be avoided by developer e.g. by listening on `ready` event, it should be his responsibility to handle this case. Otherwise, data loss will occur, because you are not handling lazy loading here.

If the issue could be caused by framework itself, I would add some caching so data/readonly mode are set despite bad timing.

engineering-this commented:
Removed workarounds for issues with iframe. To correctly test how it works you should serve `ckeditor.js` from branch `t/3115`. Also added option to run karma tests with CKEditor library loaded from url passed via command attribute.
Eg.

```
npm run test -- -u http://localhost:5000/ckeditor.js
```

Note: you might need your http server to have proper cors headers.

For sample just add script tag pointing at your `ckeditor.js` location in `sample>index.html`.

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,133 @@
+							this.$emit( 'ready', editor );
+						} );
+
+						editor.fire( 'change' );
```
>editor#change is used to emit component#input, as data can be updated by ACF, we want to emit proper event. Firing component#input manually wouldn't be consisted with how normally it is fired with an original change evt passed as a param.

It's editor initialization code, so when you are using CKEditor (without integration) you won't get `change` event when filling up an editor with data from DOM element (the one you initialize editor on). If we treat initialization `value` with the same purpose, it also shouldn't emit `change` event when mounting component.

Moreover, you are locking `undo` feature before firing `change` event and `change` event is introduced by `undo` plugin to handle creating undo snapshots, which doesn't make sense for me. It looks like workaround over the issue with synchronizing v-model between editor and vue, although why not use mentioned `$emit#input` directly?

engineering-this replied:
>Moreover, you are locking undo feature before firing change event and change event is introduced by undo plugin to handle creating undo snapshots, which doesn't make sense for me. It looks like workaround over the issue with synchronizing v-model between editor and vue, although why not use mentioned $emit#input directly?

We can use `$emit`, but undo needs to be locked so there is no initial undo step from call on `setData`.

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,133 @@
+				const data = editor.getData();
+
+				// The compatibility with the v-model and general Vue.js concept of input–like components.
+				this.$emit( 'input', data, evt, editor );
```
[`editor#change`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_editor.html#event-change) event may be called multiple times when there is actually no change inside the editor (e.g. by changing selection). Internally undo handles this case, but it may create some performance issues with Vue. Maybe we should handle this case by checking if data reallly changed (`this.value !== editor.getData()`). I have some bad feeling that Vue may do some rendering stuff when this event is called.

jacekbogdanski commented on sample/main.js:
```diff
@@ -0,0 +1,135 @@
+	router
+} );
+
+function destroyEditors( component, editorNames ) {
```
Now it just throws console errors. Please, get rid of unused / wrong code.

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,133 @@
+
+	render( createElement ) {
+		return createElement( 'div', {}, [
+			createElement( this.tagName )
```
Is there a reason for wrapping it with empty `div`? Not sure but you could probably just use:
`return createElement( this.tagName)`
and replace `$el.firstElementChild` with `$el`.

engineering-this replied:
We need two elements here. One which is component wrapper, so all component elements are grouped within single element and another element which to be passed to `CKEDITOR.replace`, let's call it `textarea` although it can be anything else.

When returning only `textarea` then editor is appended to parent component not the editor component. This breaks the whole purpose of using component based frameworks. Additionally it might lead to issues for example `v-show="false"` directive adds `display:none` to components outer element, so it probably won't hide editor which rendered outside of the component.

jacekbogdanski replied:
I see, thanks for clarifying that.

jacekbogdanski commented on tests/_test-utils/mockeditor.js:
```diff
@@ -0,0 +1,66 @@
+		return '';
+	},
+
+	setReadOnly() {
```
:thinking:

engineering-this replied:
Usually you write more verbose comments 😜

jacekbogdanski replied:
Sorry, I thought it's some leftover as this function is empty. Somehow missed filename :smile:

engineering-this replied:
At first it made writing tests a lot easier, especially before `t/3115`. Initially I wasn't sure if mocked editor is only for start, or for longer. But now when integration is nicely shaped, not much changes are expected, so I can replace mock editor with real CKEditor.

jacekbogdanski commented on sample/index.html:
```diff
@@ -0,0 +1,143 @@
+			<section>
+				<h3>Classic Editor</h3>
+				<ckeditor
+					ref="classic"
```
AFAIK `ref` is no longer needed here.

jacekbogdanski commented on sample/index.html:
```diff
@@ -0,0 +1,143 @@
+				<ckeditor
+					ref="classic"
+					type="classic"
+					tag-name="div"
```
Can't we use default `textarea`?

jacekbogdanski commented on sample/index.html:
```diff
@@ -0,0 +1,143 @@
+					ref="classic"
+					type="classic"
+					tag-name="div"
+					v-model="editors.classic.data"
```
Seems like could be cut into `editors.classic` (of course, requires model change).

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,133 @@
+	},
+
+	beforeDestroy() {
+		if ( this.instance ) {
```
`beforeDestroy` can be called before `getEditorNamespace` promise resolves.

f1ames replied:
> So, is there a way that actually editor will be initialized somewhere between beforeDestroy and destroy and in that case never destroyed properly?

Was this case already covered? I don't see any handling in the code for such case 🤔 Would it mean that editor eventually initializes even though component was destroyed?

Is it the same case as destroying after `getEditorNamespace` succeeded but before `instanceReady`? 🤔 

cc @jacekbogdanski

jacekbogdanski replied:
Overall the case is covered by our upstream changes to CKE4 where we are preventing editor initialization on destroyed DOM element, see https://github.com/ckeditor/ckeditor4/pull/3128

However, we still have https://github.com/ckeditor/ckeditor4-vue/blob/14409e98ac7fe09b601ca952cb11d8b775f9cba7/src/ckeditor.js#L49-L51 as a kind of defensive programming against this issue.

And here you have a nice unit test confirming this edge case: https://github.com/ckeditor/ckeditor4-vue/blob/14409e98ac7fe09b601ca952cb11d8b775f9cba7/tests/component.js#L211-L263

f1ames replied:
Thanks for clarification on this one 👍

jacekbogdanski commented on tests/plugin.js:
```diff
@@ -0,0 +1,44 @@
+/**
```
Isn't this test file a bit redundant? :thinking: I mean, we already testing if a component has been initialized and destroyed with `tests/integration.js` and `tests/component.js`. WDYT?

engineering-this replied:
This file tests only `plugin.install` method. When this method gets broken all or most of the other tests fails. However I'd keep this one because it's assertion might be the best clue of what wrong.

jacekbogdanski commented on tests/integration.js:
```diff
@@ -0,0 +1,152 @@
+
+import Vue from 'vue';
+import { mount } from '@vue/test-utils';
+import CKEditor from '../src/index';
```
I would rename it into `CKEditorComponent` like it has been done for `tests/component.js` file.

engineering-this replied:
Two different classes are exported.
- In `src/ckeditor.js` is `CKEditorComponent` - component
- In `src/index.js` is `CKEditor` - plugin which installs the component

In given case we are importing the plugin, so I'm keeping the `CKEditor` name to distinguish it from the component.

engineering-this commented on src/ckeditor.js:
```diff
@@ -0,0 +1,134 @@
+
+	mounted() {
+		getEditorNamespace( this.editorUrl ).then( () => {
+			const config = { ...this.config };
```
I have opposite opinion here. Spread operator is easier to read for me. But that's just personal preference.

engineering-this commented on tests/component.js:
```diff
@@ -0,0 +1,266 @@
+						sinon.assert.calledOnce( spies[ method ] );
+					} );
+
+					it( `shouldn't call "CKEDITOR.${ otherMethod }"`, () => {
```
You are right, this test can be omitted.

engineering-this commented on tests/component.js:
```diff
@@ -0,0 +1,266 @@
+		} );
+
+		describe( 'when editor type', () => {
+			describe( 'set to invalid value', () => {
```
![Screenshot 2019-06-07 at 16 00 04](https://user-images.githubusercontent.com/23366849/59109465-63176a00-893d-11e9-99c7-64e2f1c407e6.png)

My goal here was to have clean and non repetitive structure of the output. Although I see I can slightly simplify it, by moving nested `describe` into previous one after the `forEach`.

engineering-this commented:
>Did you have opportunity to check if you can actually start BrowserStack tests with some success? Not sure if this configuration has been tested or just copied from CKE5 Vue integration.

Configuration is copied. We need keys for browserstack tho. I added travis config, builds are failing because it uses released CKEditor, so until release of 4.12.0 of CKEditor it doesn't make much sense to do anything about it.

jacekbogdanski commented on tests/integration.js:
```diff
@@ -0,0 +1,138 @@
+		} );
+	} );
+
+	it( 'when component has initial data shouldn\'t produce undo steps', () => {
```
The test seems to be false positive.

engineering-this replied:
https://github.com/ckeditor/ckeditor4-vue/blob/7ba93dcf10ddba536d473795a4607d25b60abbe1/src/ckeditor.js#L64

When above line is removed test fails, so why do you find out it being false positive?

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,134 @@
+							this.$emit( 'ready', editor );
+						} );
+
+						this.$emit( 'input', newData );
```
There is no test validating synchronization between the model and editor.

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,134 @@
+
+		// Note: By the time the editor is destroyed (editor.on( 'destroy' ) called)
+		// the Vue component will not be able to emit any longer. So emitting #destroy a bit earlier.
+		this.$emit( 'destroy', this.instance );
```
I'm wondering if this event is correct :thinking: It emits `destroy` event in the context of editor instance, not the whole component. Although, I'm not sure if both should be distinguished for the end user and probably the event should be moved before destroying instance and renamed into `beforeDestroy`.

On the other hand, fixing iframe based editor should probably allow us to move this whole logic into the correct `destroy` component method. In that case, I would suggest refactoring this method to emit only `beforeDestroy` event so users can use it for some cleanup and emit correct `destroy` event when an editor is destroyed (from `component.destroy`). It also makes sense to check out if Vue is not calling similar events on its own (maybe by different API?). In that case, destroy events would be redundant.

engineering-this replied:
I'm more for removing this event at all. We can't emit real `destroy` event, replacing it with `beforeDestroy` feels a bit redundant, because whoever is using CKEditor can just use `beforeDestroy` hook of parent component. Also when needed `component.instance#destroy` event can be used.

jacekbogdanski replied:
Mostly it opens more possibilities for cleanup for users. However, as you pointed out there is some workaround for it, so let's wait if the feature is implicit requested by the community.

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,134 @@
+			}
+		},
+
+		readOnly( val ) {
```
Seems like no unit test coverage here.

engineering-this replied:
Is there any particular case you'd like to cover? I've added plenty of test for `readOnly`, ones that directly check commented setter are here:
https://github.com/ckeditor/ckeditor4-vue/blob/7ba93dcf10ddba536d473795a4607d25b60abbe1/tests/component.js#L234-L244

jacekbogdanski replied:
Removing this method from class results in the same failing tests as before, so looks like no test coverage or tests are failing positive.

f1ames replied:
Was it already covered cc @jacekbogdanski?

jacekbogdanski replied:
You can just remove https://github.com/ckeditor/ckeditor4-vue/blob/14409e98ac7fe09b601ca952cb11d8b775f9cba7/src/ckeditor.js#L106-L108 from code to see it yourself, but yeah, couple of tests are dying now :skull: 
![image](https://user-images.githubusercontent.com/10279586/72555610-3517e200-389d-11ea-857d-66497dbb6fab.png)

engineering-this commented on tests/component.js:
```diff
@@ -0,0 +1,281 @@
+		sandbox = sinon.createSandbox();
+
+		component.$once( 'ready', () => {
+			done();
```
We can't pass `done` as direct listener to an event, it'd break tests. `done` accepts optional argument, an editor instance would be passed which is not expected by `done` function.

engineering-this commented on sample/index.html:
```diff
@@ -0,0 +1,139 @@
+	</style>
+</head>
+<body>
+	<script src="http://localhost:5000/ckeditor.js"></script>
```
Didn't mean to commit that 🙄

engineering-this commented:
## Destroying component before `getEditorNamespace` resolves

Fixing it was trivial, writing test which covers it was a bit more challenging. There is no way to mock/stub ESM imports, at least I didn't find any without extra deps. Fortunately method that had to be mocked is `getEditorNamespace` which is only for our internal use, I've extended it with debug tools, that allows to easily replace the method.

## Coverage

After inspecting what coverage shows I fixed some tests, so now we have 100% code coverage.

## Browserstack

Permission Denied thrown by Edge was fixed in the Editor with recent commits to `t/3115`.

```
Safari 12.1.1 (Mac OS X 10.14.5): Executed 52 of 52 SUCCESS (13.748 secs / 0.254 secs)
Edge 18.18362.0 (Windows 10.0.0): Executed 52 of 52 SUCCESS (17.376 secs / 0.283 secs)
TOTAL: 104 SUCCESS
```

## IE11 and lower

[`@babel/preset-env`](https://babeljs.io/docs/en/babel-preset-env) allow setting target browsers, and it will transpile code and load needed polyfils from `core-js` to work in target envs. However, some tests fails because of missing `object.assign`, even when implicitly declaring missing polyfill in webpack config, it doesn't help. Looks like some issue with webpack/karma. I don't think it's worth to spend more time on it right now. We could extract testing IEs to follow-up issue.

jacekbogdanski commented on src/utils/geteditornamespace.js:
```diff
@@ -0,0 +1,52 @@
+
+let promise, mock;
+
+export function mockMethod( callback ) {
```
Oh no, we cannot do that :smile: If you really need to hack into method internals, make them public. So instead of making `mock` object available, do it with `loadScript` library. You could just:

```javascript
getEditorNamespace.scriptLoader = loadScript;
```

making this property open for mocking. Also, I recommend promising `loadScript` before it's even attached as a static member for easier mocking later.

monas commented:
I was trying to run the tests and to see if I could help in any way. I am pretty new to Vue testing so for me it would be a nice opportunity to learn new stuff. Also keep in mind my lack of experience and check twice what I am saying. 
What I have discovered is that in utils test there is something wrong with beforeEach/afterEach blocks. I think on of those are not needed. At least for me after commenting this line (https://github.com/ckeditor/ckeditor4-vue/blob/1925e8ef501c00f1716f0594710121177f0af1c1/tests/utils.js#L17) the utils tests are passing.

engineering-this commented:
Just a reminder. You need to run tests with `ckeditor.js` from `t\3115` branch. Otherwise they will fail.

e.g.
```
npm run tests -- -u http://localhost:5000/ckeditor.js
```

>Could you also make README.md more consistent with react and angular integrations? You could use the same pattern with changes related to the current integration.

Not really part of MVP, we can extract it to separate ticket.

jacekbogdanski commented:
Taking over development.

jacekbogdanski commented:
I changed the target branch so we won't accidentally merge it to the `master` as README.md file is written as production-ready, so it may be incorrectly interpreted by the community. Let's merge `development` into `master` once work is done (including docs).

Comandeer commented on tests/component.js:
```diff
@@ -0,0 +1,351 @@
+		spies.inline = sinon.spy( CKEDITOR, 'inline' );
+
+		wrapper = createComponent( props );
+		component = wrapper.vm;
```
TBH I'm not keen on doing such things outside of the test. IMO `beforeEach` should only create the environment to run the test. However the component being the subject of the test should be created directly inside the test.

The creation of the component can be exported to the function.

jacekbogdanski replied:
I don't have a real issue with the code (note that it wasn't written by me) where SUT is created in setup function. It seems like a pretty popular practice (see CKEditor4 tests where an editor is also created in setup functions, we just don't see it in unit tests) and reduces the complexity of the unit tests. I agree it may be a bit more convenient to have SUT building function, but it's not worth refactoring the whole test file at this moment.

Comandeer commented on tests/component.js:
```diff
@@ -0,0 +1,351 @@
+
+	beforeEach( done => {
+		spies.replace = sinon.spy( CKEDITOR, 'replace' );
+		spies.inline = sinon.spy( CKEDITOR, 'inline' );
```
Why aren't these two spies inside the sandbox? I do not also see the need to spy on these two methods in every test.

jacekbogdanski replied:
Good point about the sandbox :+1: However, it's just simpler to spy on them with topmost setup function, so we don't duplicate these two lines. Also, creates nice testing API with sandbox.

Comandeer commented on tests/component.js:
```diff
@@ -0,0 +1,351 @@
+	afterEach( () => {
+		skipReady = false;
+
+		for ( const key in spies ) {
```
Do we really need names for spies? I'd use array for `spies` instead.

jacekbogdanski replied:
Note how spies are used in this test file. They are referenced by name by multiple test steps and then restored after each test.

Comandeer commented on tests/component.js:
```diff
@@ -0,0 +1,351 @@
+
+					if ( defaultValueRegex ) {
+						expect( component[ property ] ).to.match( defaultValueRegex );
+					}
```
Personally I'd simplify it by accepting only regexes. This way string values would become `/^string$/`.

Comandeer commented on tests/component.js:
```diff
@@ -0,0 +1,351 @@
+			} );
+
+			it( `should emit "${ evtName }"`, () => {
+				sinon.assert.calledOnce( spy );
```
Most of the test logic resided in `beforeEach` hook. I'd move it inside `it`.

jacekbogdanski replied:
Removing logic from `beforeEach` where we are firing the first input event may result with false-positive test for `input` event later (see a second `it`). Could you tell which parts you would move?

jacekbogdanski replied:
Actually, it seems like there is no need to treat it any special as component value is not updated by `input` event (only spied). Looks like this test needs some refactoring thou. I will split it into 2 different test cases, as it's unreadable in the current format.

jacekbogdanski commented on .eslintrc.js:
```diff
@@ -0,0 +1,15 @@
+
+// Note: The ESLint configuration is mandatory for vue-cli.
+module.exports = {
+	'extends': 'ckeditor5',
```
Yes, but the one from CKEditor5 seems to be better customized for ES6 and also is still maintained.

jacekbogdanski commented on .gitattributes:
```diff
@@ -0,0 +1,18 @@
+*.txt		text
+*.md		text
+
+*.png		-text
```
Do you mean `*.png    -text`? Note `-` sign, it's unsetting path operation, not treating `png` as a text. Refer to https://git-scm.com/docs/gitattributes how `.gitattributes` work.

Comandeer replied:
I didn't see that `-` 😞 I probably treated it as another tab character. Now it makes perfect sense 👍

jacekbogdanski commented on .gitignore:
```diff
@@ -1,37 +1,2 @@
-.DS_Store
-Thumbs.db
+coverage/
+node_modules/
```
I'm not sure how publishing will look like, however, attaching `dist` folder is important if someone is adding library via git path. I believe it won't change for `npm` publishing, so `dist` package should be also pushed into a repository and updated when needed. The same has been done for CKEditor5 Vue integration - https://github.com/ckeditor/ckeditor5-vue and is done for most popular Vue plugins, e.g. see `vue-router` - https://github.com/vuejs/vue-router

Comandeer replied:
However our React integration, `ckeditor4-react`, does not include `dist` directory – even for tags. Probably we can rethink the whole distribution system. I'm for pushing `dist` directory only for tags (similar to e.g. `rollup-plugin-babel-minify`). `vue-router` has always `dist` directory, but it seems to be updated only for tags – so it won't be helpful for adding library via git.

jacekbogdanski replied:
>  I'm for pushing dist directory only for tags (similar to e.g. rollup-plugin-babel-minify)

What do you mean about tags? :thinking: 

I'm fine for removing `dist` folder for dev version as using git path is not necessary a valid way to include a plugin (especially for production version), however, how would be `dist` folder published by npm? By adding it to `files` option and building + publishing on every release?

Comandeer replied:
> What do you mean about tags?

https://github.com/Comandeer/rollup-plugin-babel-minify/blob/90ca4099d88387fce7eb39f8364a0ce1036250f5/package.json#L18-L19

> how would be dist folder published by npm? By adding it to files option and building + publishing on every release?

Yes.

jacekbogdanski replied:
Ok, it looks like the best option for me :+1:  Also this 'way' of versioning is described in npm docs (https://docs.npmjs.com/cli/version) if someone needs more depth documentation.

f1ames replied:
I'm not sure if we need `dist/` in the main repo? What do you mean by:

> if someone is adding library via git path

is it a case when one just clones the repo and uses it? I think in this case `dist/` should be simply build by the person doing that as he is using dev (not production/dist) version of the integration. And if you are talking about using repo as node module I would be more for adding build step as a postinstall step.
And as for tags - since tags never change one also may build `dist/` after cloning the tag (I assume it's one command, right?).

jacekbogdanski replied:
> is it a case when one just clones the repo and uses it?

no, it's a case when you are using github as a npm package path, e.g.

```
dependencies: {
  "ckeditor4-vue": "https://github.com/ckeditor/ckeditor4-vue.git#1.0.0"
}
```
but as I understand, versioning proposed by @Comandeer will take care of keeping `dist` folders in git tag.

`dist` folder is required for npm to work, otherwise requiring plugin from node won't load correctly (e.g. `import CKEditor4 from 'ckeditor4-vue` will throw error).

f1ames replied:
Yes, I think that's what @Comandeer had in mind. Can't we make building (so creating `dist/`) a post-install step then? WDYT @Comandeer @jacekbogdanski 
Not sure what would be the benefit apart from not having `dist/` folder (which was the main discussion point here) :thinking:

jacekbogdanski replied:
A simple post-install script should also probably work here, so `dist` is produced only for npm installation. :+1: from me (if it will work). Both options are fine - with post-install we are losing dist tags versioning when using gitpath, but it's probably not really important here.

I don't have a really hard opinion on which one method is better, but considering that @f1ames one seems less complicated, I'm probably slightly more for it. @Comandeer do you have some arguments which could help choose using versioning script with git tags over simplified @f1ames idea?

Comandeer replied:
Hmmm, so user installing package from npm will get `dist/`, but the package will rebuild it due to `postinstall` script?

f1ames replied:
Hmmm, good point, I was thinking only about rebuilding in postinstall when github repo is used as NPM package, not a dist package - however this again will add some complexity to the process as we will need different flow for dist version (no postinstall step) :thinking:

jacekbogdanski replied:
So, @Comandeer way than? It seems legit.

jacekbogdanski commented on package.json:
```diff
@@ -0,0 +1,60 @@
+  "homepage": "https://github.com/ckeditor/ckeditor4-vue",
+  "dependencies": {
+    "load-script": "^1.0.0"
+  }
```
Not sure about that. Plugin don't have any special references to the Vue except unit tests and dev samples. Could you tell more why you think it should be moved into `peerDependency`?

jacekbogdanski commented on samples/main.js:
```diff
@@ -0,0 +1,95 @@
+
+const router = new VueRouter( { routes } );
+
+/* eslint no-new: 0 */
```
This one
```javascript
new Vue( { // eslint-disable-line no-new
  el: '#app',
   router
} );
```
is not really readable. Do we really discussing which `eslint` rule format is better? :smile:

jacekbogdanski commented on src/utils/geteditornamespace.js:
```diff
@@ -0,0 +1,41 @@
+	return promise;
+}
+
+getEditorNamespace.scriptLoader = editorURL => new Promise( ( scriptResolve, scriptReject ) => {
```
This function is used in unit tests, see https://github.com/ckeditor/ckeditor4-vue/blob/d80e9f27d366262fba16a0244291b888ff99c947/tests/component.js#L236

jacekbogdanski commented on src/utils/geteditornamespace.js:
```diff
@@ -0,0 +1,41 @@
+			scriptReject( new Error( 'Script loaded from editorUrl doesn\'t provide CKEDITOR namespace.' ) );
+		} else {
+			scriptResolve( CKEDITOR );
+		}
```
Rejecting promise doesn't magically prevent calling resolving method. It will work due to internal Promise behavior, but why not write it correctly, so anything going to resolve method won't be executed?

Comandeer commented on package.json:
```diff
@@ -0,0 +1,66 @@
+	"homepage": "https://github.com/ckeditor/ckeditor4-vue",
+	"dependencies": {
+		"load-script": "^1.0.0"
+	}
```
I know that it's not much important, but… can we move this section to others dependencies? It's triggering me.

jacekbogdanski replied:
Not sure if I understand :thinking: Where would you like to move this dependency specifically?

Comandeer replied:
Not the dependency, but whole section, to group it with other deps:

```json
"devDependencies": {},
"peerDependencies": {},
"dependencies": {}
```

Comandeer commented on webpack.config.js:
```diff
@@ -0,0 +1,66 @@
+		rules: [
+			{
+				test: /\.js$/,
+				loader: 'babel-loader',
```
We should probably specify which preset should be used and which browsers we target.

jacekbogdanski replied:
Could you give a hand with that?

Comandeer replied:
We're already doing it in React, you can try if settings provided there works: https://github.com/ckeditor/ckeditor4-react/blob/9e5955f1c81c8670e460e2cf69352eae6e324fad/webpack.config.js#L78-L97

jacekbogdanski commented:
Yes, I already extracted IE11 issue here - https://github.com/ckeditor/ckeditor4-vue/issues/3

I'm for doing IE11 research as a followup to the current version as from our experience IE11 may take some time to handle. WDYT? Also, cc @f1ames do you see IE11 support obligatory for the first beta?

Comandeer commented:
> do you see IE11 support obligatory for the first beta?

For me it's a must have TBH.

Comandeer commented:
It seems that after a little polishing of samples code, support for IE11 works properly 👍. @jacekbogdanski, can you verify?

jacekbogdanski commented:
Works good, thanks for help :+1: I customized a bit webpack config so now we are building additional, separate `dist/legacy.js` with IE11 support. Otherwise, we would end up with a much bigger file bundle (4KiB vs 22KiB) with IE11 support which is not often required.

jacekbogdanski commented on package.json:
```diff
@@ -0,0 +1,67 @@
+	},
+	"scripts": {
+		"start": "npm run build && live-server --open=samples/index.html",
+		"build": "webpack --mode production && webpack --mode production --env.legacy='true'",
```
Didn't know about this option - indeed looks much more elegant, thanks! :+1:

jacekbogdanski commented on src/ckeditor.js:
```diff
@@ -0,0 +1,139 @@
+
+				if ( undo ) {
+					undo.lock();
+				}
```
Not really, as it seems like in that case callback will be never called due to missing `dataReady` event (an internal option also disables flow responsible for this event to be called). Although, I find out firing `lockSnapshot/unlockSnapshot` more elegant than checking against available undo plugin.

jacekbogdanski commented:
I updated the sample description, so now it correctly tells about the required textarea blur.

> What's more, selection becomes blurry when you switch right from textarea to the editor: caret position is reset (in IE11) or selection is totally removed (Chrome).

I don't think it's necessary to spend more time introducing throttling or other mechanisms to delay setting editor data. An example would be more complicated and the main feature (2-way binding via v-model) hidden under the fix. Note it's dev example also, so it's more about implementation show off than UX.

f1ames commented on src/ckeditor.js:
```diff
@@ -0,0 +1,134 @@
+		type: {
+			type: String,
+			default: 'classic',
+			validator: type => [ 'classic', 'inline' ].includes( type )
```
What about `divarea` editor? 🤔

jacekbogdanski replied:
`divarea` is just a plugin, why it should be considered as a type?

f1ames replied:
Yes, you are right that `divarea` is only the plugin and can be simply added so it makes sense. I'm not sure TBH for some reason we have `classic`, `divarea`, `inline` in Angular integration but only `classic` and `inline` in React.
I think in Angular, there was an issue with Classic one (iframed) and we went with divarea plugin to temporarily substitute it 🤔 

Looking at our docs, the type is Classic/Inline so we should stick with it to be consistent:

![image](https://user-images.githubusercontent.com/1061942/72592967-18fb5b80-3904-11ea-8ea2-630ed5b561b9.png)

Which means two types are fine here 👍

f1ames commented on src/ckeditor.js:
```diff
@@ -0,0 +1,134 @@
+
+			editor.on( 'blur', evt => {
+				this.$emit( 'blur', evt, editor );
+			} );
```
I see we don't have too many events exposed here, but maybe let's wait for feedback and see if these are needed.

jacekbogdanski replied:
It's mostly the same set of events as for angular integration. Devs still can just use `config#on` for less common events or even access editor instance directly from the component.

Comandeer commented on .travis.yml:
```diff
@@ -0,0 +1,17 @@
+sudo: required
+dist: trusty
+addons:
```
Actually, aren't our integrations a good place to experiment with GitHub Actions?

f1ames replied:
A perfect place I would say, but I would be for doing it as a separate issue to not block beta release. WDYT?

f1ames replied:
@Comandeer could you extract it to a separate issue?

f1ames replied:
@Comandeer do you think we could easily set up CD with GH Actions? Like automatically publishing tags and e.g. tagging based on commit messages or something like that? So releasing new version will simply require pushing a single commit to `master`?

jacekbogdanski commented on README.md:
```diff
@@ -1,8 +1,95 @@
 Official [CKEditor 4](https://ckeditor.com/ckeditor-4/) WYSIWYG editor component for Vue.
 
-### License
+We are looking forward to your feedback! You can report any issues, ideas or feature requests on the [integration issues page](https://github.com/ckeditor/ckeditor4-vue/issues/new).
```
I don't either.

f1ames replied:
> I don't either.

Not sure if I should be happy or not 😝 But beta is needed indeed 👍

jacekbogdanski commented on package.json:
```diff
@@ -0,0 +1,67 @@
+		"develop": "webpack --mode development --watch",
+		"test": "karma start",
+		"preversion": "npm test",
+		"version": "npm run build && git add -f dist/",
```
Running `npm version` also runs `preversion` and `postversion` scrips: https://github.com/ckeditor/ckeditor4-vue/blob/165df1607e53f3cdbf4741704a537ca6e9a43213/package.json#L53-L55

`npm version` also takes care of creating version tag and committing it to a repository, see https://docs.npmjs.com/cli/version#description

f1ames replied:
Ok, I thought that we somehow redefined `version` so it will only run what is declared here. The docs says it will run default `npm version` and then our custom part so 👍 

We should utilize it in other integrations too ⚙️

jacekbogdanski commented on karma.conf.js:
```diff
@@ -0,0 +1,188 @@
+				os: 'OS X',
+				os_version: 'High Sierra',
+				browser: 'safari'
+			}
```
I tried to add IE11 support by extending also karma config with IE11 presets, but still it somehow fails due to missing `object.assign` polyfill. I will extract it as a followup ticket as suggested.

jacekbogdanski commented:
There is no problem with retargeting the current upstream right now. I suppose that going with `master` and `stable` branches, this PR should target stable now? `development` branch has been only added to prevent accidental merging this PR into `master` with README which could claim that we are done with beta development before publishing docs / making latest corrections.
